spec/unit/puppet_x/libkv/file_plugin_spec.rb

```diff
@@ -0,0 +1,502 @@
+        key_dir = File.join(@root_path, 'production')
+        FileUtils.mkdir_p(key_dir)
+        result = @plugin.list('production')
+        expect( result[:result] ).to eq({})
+        expect( result[:err_msg] ).to be_nil
+      end
+
+      it 'should return full list of key/value pairs in :result when key folder content is accessible' do
+        expected = {
+          'production/key1' => 'value for key1',
+          'production/key2' => 'value for key2',
+          'production/key3' => 'value for key3'
+        }
+        expected.each { |key,value| @plugin.put(key, value) }
+        result = @plugin.list('production')
+        expect( result[:result] ).to eq(expected)
+        expect( result[:err_msg] ).to be_nil
+      end
+
+      it 'should return partial list of key/value pairs in :result when some key folder content is not accessible' do
+        expected = {
+          'production/key1' => 'value for key1',
+          'production/key3' => 'value for key3'
+        }
+        expected.each { |key,value| @plugin.put(key, value) }
+
+        # create a file for 'production/key2', but make it inaccessible via a lock
+        locked_key_file_operation(@root_path, 'production/key2', 'value for key2') do
+          puts "     >> Executing plugin list() for 'production'"
+          result = @plugin.list('production')
+          expect( result[:result] ).to eq(expected)
+          expect( result[:err_msg] ).to be_nil
+        end
+
+      end
+
+      it 'should return an unset :result  and an :err_msg when key folder does not exist or is inaccessible' do
+        result = @plugin.list('production')
+        expect( result[:result] ).to be_nil
+        expect( result[:err_msg] ).to match(/Key folder not found/)
+      end
+    end
+
+    describe 'name' do
+      it 'should return configured name' do
+        expect( @plugin.name ).to eq 'file/test'
+      end
+    end
+
+    # using plugin's get() in this test, because it has already been
+    # fully tested
+    describe 'put' do
+      it 'should return :result=true when the key file does not exist for a simple key' do
+        key = 'key1'
+        value = 'value for key1'
+        result = @plugin.put(key, value)
+        expect( result[:result] ).to be true
+        expect( result[:err_msg] ).to be_nil
+        expect( @plugin.get(key)[:result] ).to eq value
+        key_file = File.join(@root_path, key)
+        expect( File.stat(key_file).mode & 0777 ).to eq 0640
+      end
+
+      it 'should return :result=true when the key file does not exist for a complex key' do
+        key = 'production/gen_passwd/key1'
+        value = 'value for key1'
+        result = @plugin.put(key, value)
+        expect( result[:result] ).to be true
+        expect( result[:err_msg] ).to be_nil
+        expect( @plugin.get(key)[:result] ).to eq value
+        key_file = File.join(@root_path, key)
+        expect( File.stat(key_file).mode & 0777 ).to eq 0640
```

Comment:
judy johnson 5:03 PM

error message if your umask is not set up as expected:

1) libkv file plugin anonymous class public API put should return :result=true when the key file does not exist for a simple key
    Failure/Error: expect( File.stat(key_file).mode & 0777 ).to eq 0640
      expected: 416
           got: 384
      (compared using ==)
    # ./spec/unit/puppet_x/libkv/file_plugin_spec.rb:438:in `block (4 levels) in <top (required)>'